Mathlib/Analysis/SpecialFunctions/BinEntropy.lean

```diff
@@ -0,0 +1,508 @@
+    qaryEntropy q p = p * log (q - 1) + binaryEntropy p := by
+  unfold binaryEntropy qaryEntropy
+  simp only [show ((2:ℕ) - (1:ℝ)).log = 0 by norm_num]
+  ring
+
+lemma qaryEntropy_eq_log_mul_add_binaryEntropy' {q : ℕ} :
+    qaryEntropy q = (fun p ↦ p * log (q - 1)) + binaryEntropy := by
+  ext
+  simp only [Pi.add_apply, qaryEntropy_eq_log_mul_add_binaryEntropy]
+
+lemma binaryEntropy_pos {p : ℝ} (pgt0 : 0 < p) (plt1 : p < 1) : 0 < binaryEntropy p := by
+  simp only [binaryEntropy_eq']
+  have pos_sum_pos_pos (a b : ℝ) (ha : 0 ≤ a) (hb : b < 0) : 0 < a - b := by linarith
+  refine pos_sum_pos_pos (-p * log p) ((1 - p) * log (1 - p)) ?_ ?_
+  · rw [show -p * log p = p * (-log p) by ring]
+    exact (Real.mul_pos (by linarith) (by linarith [log_neg pgt0 plt1])).le
+  · exact mul_neg_of_pos_of_neg (by linarith) (log_neg (by linarith) (by linarith))
+
+lemma qaryEntropy_pos {q : ℕ} {p : ℝ} (pgt0 : 0 < p) (plt1 : p < 1) : 0 < qaryEntropy q p := by
+  unfold qaryEntropy
+  have p_q_log_nonneg : 0 ≤ p * ((q : ℝ) - 1).log := by
+    rw [mul_nonneg_iff_of_pos_left pgt0, show q - (1 : ℝ) = (q - 1 : ℤ) by norm_cast]
+    exact Real.log_intCast_nonneg _
+  have rest_is_pos : 0 < -(p * p.log) - (1 - p) * (1 - p).log := by
+    simp only [← neg_mul, ← binaryEntropy_eq']
+    exact binaryEntropy_pos pgt0 plt1
+  linarith
+
+/- Outside usual range of `binaryEntropy`. This is due to `log x = log |x|` -/
+lemma binaryEntropy_neg_of_neg {p : ℝ} (hp : p < 0) : binaryEntropy p < 0 := by
+  simp only [binaryEntropy_eq']
+  suffices -p * log p < (1-p) * log (1-p) by linarith
+  by_cases hp' : p < -1
+  · have : log p < log (1 - p) := by
+      rw [← log_neg_eq_log]
+      exact log_lt_log (Left.neg_pos_iff.mpr hp) (by linarith)
+    nlinarith [log_pos_of_lt_neg_one hp']
+  · have : -p * log p ≤ 0 := by
+      wlog h : -1 < p
+      · simp only [show p = -1 by linarith, log_neg_eq_log, log_one, le_refl, mul_zero]
+      · nlinarith [log_neg_of_lt_zero hp h]
+    nlinarith [(log_pos (by linarith) : 0 < log (1 - p))]
+
+/- Outside usual range of `binaryEntropy`. This is due to `log x = log |x|` -/
+lemma binaryEntropy_neg_of_gt_one {p : ℝ} (hp : 1 < p) : binaryEntropy p < 0 := by
+  let x := p - 2⁻¹
+  rw [show p = 2⁻¹ + x by ring, binaryEntropy_two_inv_add]
+  have : 2⁻¹ - x < 0 := by ring_nf; linarith
+  exact binaryEntropy_neg_of_neg this
+
+lemma binaryEntropy_zero_iff_zero_or_one {p : ℝ} : binaryEntropy p = 0 ↔ p = 0 ∨ p = 1 := by
+  constructor <;> intro h
+  · by_cases plt0 : p < 0
+    · linarith [binaryEntropy_neg_of_neg plt0]
+    · by_cases pgt1 : p > 1
+      · linarith [binaryEntropy_neg_of_gt_one pgt1]
+      · by_cases pz : p = 0
+        · left; assumption
+        · by_cases pone : p = 1
+          · right; assumption
+          · have : 0 < binaryEntropy p := by
+              apply binaryEntropy_pos (Ne.lt_of_le (fun a ↦ pz a.symm) (le_of_not_lt plt0))
+              exact Ne.lt_of_le pone (le_of_not_lt pgt1)
+            linarith
```

Comment:
No point indenting the main part of a proof. Also please use `obtain` along with `rfl` and `eq_or_ne` for shorter proofs which are still readable.
```suggestion
    · by_cases pgt1 : p > 1
      · linarith [binaryEntropy_neg_of_gt_one pgt1]
      obtain rfl | pz := eq_or_ne p 0
      · simp
      obtain rfl | pone := eq_or_ne p 1
      · simp
      have : 0 < binaryEntropy p := by
        apply binaryEntropy_pos (pz.lt_of_le (le_of_not_lt plt0))
        exact pone.lt_of_le (le_of_not_lt pgt1)
      linarith
```

Reply:
In fact, you are doing way too much work with the order disjunctions here. Here is a shorter proof
```suggestion
  · contrapose! h
    obtain hp₀ | hp₀ := h.1.lt_or_lt
    · exact (binaryEntropy_neg_of_neg hp₀).ne
    obtain hp₁ | hp₁ := h.2.lt_or_lt.symm
    · exact (binaryEntropy_neg_of_gt_one hp₁).ne
    · exact (binaryEntropy_pos hp₀ hp₁).ne'
```